### PR TITLE
[DEV-2686] JWT TTL and management

### PR DESCRIPTION
**Pull request by morestatic**
https://tracker.rport.io/issue/DEV-2686/JWT-ttl-management

### CODE REVIEW

File: server/api/session/cache.go [hunk 1]
```diff
@@ -94,3 +93,90 @@ func (p *Cache) getFromCache(token string) (*APISession, error) {
 
 	return existing, nil
 }
+
+func (p *Cache) GetAllByUser(ctx context.Context, username string) (sessions []*APISession, err error) {
+	count := p.cache.ItemCount()
+	sessions = make([]*APISession, 0, count)
```
Comment (domenipavec) [inline server/api/session/cache.go:99]: This is allocation is too big. For a system with large number of users this allocation could be many times too big. Since actual number if unknown I think it's better to just avoid preallocation.

File: server/api/session/cache.go [hunk 2]
```diff
@@ -94,3 +93,90 @@ func (p *Cache) getFromCache(token string) (*APISession, error) {
...
+		}
+	}
+
+	return nil
+}
+
+func (p *Cache) deleteUserSessionFromStorage(ctx context.Context, username string, sessionID int64) (err error) {
+	err = p.storage.DeleteByUser(ctx, username, sessionID)
+	if err != nil {
+		return fmt.Errorf("unable to delete session from cache: %w", err)
```
Comment (domenipavec) [inline server/api/session/cache.go:159]: This is deleting from storage not cache.

File: server/api/session/cache.go [hunk 3]
```diff
@@ -94,3 +93,90 @@ func (p *Cache) getFromCache(token string) (*APISession, error) {
...
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Cache) deleteUserSessionsFromStorage(ctx context.Context, username string) (err error) {
+	err = p.storage.DeleteAllByUser(ctx, username)
+	if err != nil {
+		return fmt.Errorf("unable to delete sessions from cache: %w", err)
```
Comment (domenipavec) [inline server/api/session/cache.go:179]: Same

File: server/api/session/cache.go [hunk 4]
```diff
@@ -94,3 +93,90 @@ func (p *Cache) getFromCache(token string) (*APISession, error) {
...
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Cache) deleteUserSessionsFromCache(username string, sessionID int64) (err error) {
+	// Items() returns a copy of the underlying unexpired cache items and Delete
+	// won't error if item not found. this should be thread safe.
+	for _, item := range p.cache.Items() {
```
Comment (domenipavec) [inline server/api/session/cache.go:133]: Looping over items for single item deletion isn't ideal.
You have a problem here because you have 2 fields used as primary keys (sessionID and token). I think you should use `sessionID` for keying the cache and encrypt it in jwt tokens. That way tokens can be converted to session ids.

File: server/config.go [hunk 5]
```diff
@@ -55,6 +56,7 @@ type APIConfig struct {
 	UserLoginWait         float32 `mapstructure:"user_login_wait"`
 	MaxFailedLogin        int     `mapstructure:"max_failed_login"`
 	BanTime               int     `mapstructure:"ban_time"`
+	MaxTokenLifeTime      int     `mapstructure:"max_token_lifetime"`
```
Comment (domenipavec) [inline server/config.go:59]: This should either be `time.Duration` or have a suffix specifying the unit.

File: db/migration/api_sessions/sql/002_extended_session_tracking.up.sql [hunk 6]
```diff
@@ -0,0 +1,22 @@
+-- TODO: currently all existing sessions will be cleared. is this ok?
```
[the remark this answers is not in the document]
Comment (thorstenkramm) [reply db/migration/api_sessions/sql/002_extended_session_tracking.up.sql:1]: If all sessions are killed during the db migration, this is acceptable. This is done only once. I'll add a warning to the release notes.

File: server/config.go (hunk 5 above)
Comment (morestatic) [reply server/config.go]: ok. added time unit suffix to field name.

File: server/api/session/cache.go (hunk 2 above)
Comment (morestatic) [reply server/api/session/cache.go]: well spotted. done.

File: server/api/session/cache.go (hunk 3 above)
Comment (morestatic) [reply server/api/session/cache.go]: done.

File: server/api/session/cache.go (hunk 1 above)
Comment (morestatic) [reply server/api/session/cache.go]: yep. removed the preallocation.

File: server/api_handler_login_test.go [hunk 7]
```diff
@@ -227,6 +228,147 @@ func TestWrapWithAuthMiddleware(t *testing.T) {
...
+				username := tc.Username
+				if tc.BearerUsername != "" {
+					username = tc.BearerUsername
+				}
+				sessions, err := al.apiSessions.GetAllByUser(ctx, username)
+				require.NoError(t, err)
+
+				if tc.ExpectedSession {
+					require.NotNil(t, sessions)
+					require.Less(t, 0, len(sessions))
```
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api_handler_login_test.go]: ok. made the switch. done.

File: server/api_handler_users_test.go [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+		bannedUsers: security.NewBanList(0),
+		apiSessions: sessionCache,
+		userService: users.NewAPIService(users.NewStaticProvider([]*users.User{adminUser}), false),
+	}
+	al.initRouter()
+
+	return al, adminUser
+}
+
+func basicAuth(username, password string) string {
```
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api_handler_users_test.go]: agreed. done.

File: server/api/session/session.go [hunk 9]
```diff
@@ -3,18 +3,40 @@ package session
...
 	"context"
 	"time"
+
+	"github.com/patrickmn/go-cache"
 )
 
 type APISession struct {
-	Token     string    `db:"token"`
-	ExpiresAt time.Time `db:"expires_at"`
+	SessionID    int64     `db:"session_id"`
```
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api/session/session.go]: yes, i just forgot to put these in. done.

File: server/api_handler_users.go [hunk 10]
```diff
@@ -123,3 +132,87 @@ func (al *APIListener) handleDeleteUsersTotP(w http.ResponseWriter, req *http.Re
...
+	err = al.apiSessions.DeleteByUser(ctx, userID, sessID)
+	if err != nil {
+		titleMsg := fmt.Sprintf("unable to delete session \"%s\" for user \"%s\"", sessionID, userID)
+		al.jsonErrorResponseWithDetail(w, http.StatusInternalServerError, "", titleMsg, err.Error())
+		return
+	}
+
+	al.auditLog.Entry(auditlog.ApplicationAuthAPISession, auditlog.ActionDelete).
+		WithHTTPRequest(req).
+		WithID(userID).
```
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api_handler_users.go:189]: i did think a little about this and felt (without digging too deep) that the session ID might have a short lifetime and could be difficult to discover and that the user ID was a better key in the context of the audit trail. and the session ID would still be in the logged request. but i also didn't spend too long thinking about it. any thoughts @thorstenkramm?

File: server/api_handler_users.go [hunk 11]
```diff
@@ -123,3 +132,87 @@ func (al *APIListener) handleDeleteUsersTotP(w http.ResponseWriter, req *http.Re
...
+	ctx := req.Context()
+
+	err := al.apiSessions.DeleteAllByUser(ctx, userID)
+	if err != nil {
+		titleMsg := fmt.Sprintf("unable to delete all sessions for user \"%s\"", userID)
+		al.jsonErrorResponseWithDetail(w, http.StatusInternalServerError, "", titleMsg, err.Error())
+		return
+	}
+
+	al.auditLog.Entry(auditlog.ApplicationAuthAPISession, auditlog.ActionDelete).
```
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api_handler_users.go:212]: yes, also thought a little about this. but wasn't too sure how important tracking of the individual session removal was, especially as go-cache will be expiring and removing them without any audit trails. i can understand a point of view that says explicit removal is significant and should be logged. again @thorstenkramm any thoughts?

File: server/api/session/sqlite.go [hunk 12]
```diff
@@ -100,3 +110,30 @@ func (p *SqliteProvider) DeleteExpired(ctx context.Context) error {
...
+		username,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *SqliteProvider) DeleteByUser(ctx context.Context, username string, sessionID int64) (err error) {
```
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api/session/sqlite.go:127]: agreed. renamed to DeleteByID. done.

File: server/api.go [hunk 13]
```diff
@@ -7,28 +7,29 @@ import (
...
-	routeParamClientID       = "client_id"
-	routeParamClientAuthID   = "client_auth_id"
-	routeParamUserID         = "user_id"
-	routeParamJobID          = "job_id"
-	routeParamGroupID        = "group_id"
-	routeParamVaultValueID   = "vault_value_id"
-	routeParamScriptValueID  = "script_value_id"
-	routeParamCommandValueID = "command_value_id"
-	routeParamGraphName      = "graph_name"
+	// routeParamClientID       = "client_id"
```
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api.go]: done.

File: db/migration/api_sessions/sql/002_extended_session_tracking.up.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (morestatic) [reply db/migration/api_sessions/sql/002_extended_session_tracking.up.sql:1]: i think sqlite is particularly constrained with respect to altering tables. i think there are hacks but they can have their consequences and dropping and recreating the table seems to be recommended approach.

File: server/api/session/session.go [hunk 14]
```diff
@@ -3,18 +3,40 @@ package session
...
+	Token        string    `db:"token"`
```
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api/session/session.go]: yes, i know what you mean. might be a good time to move to a more session based approach overall and a good chance to reduce the visibility/exposure of the jwt tokens in general? @thorstenkramm any thoughts from your perspective?

File: server/api_handler_logout.go [hunk 15]
```diff
@@ -17,26 +18,37 @@ func (al *APIListener) handleDeleteLogout(w http.ResponseWriter, req *http.Reque
...
 
-	token, err := al.parseToken(tokenStr)
+	tokenCtx, err := bearer.ParseToken(tokenStr, al.config.API.JWTSecret)
 	if err != nil {
 		al.jsonErrorResponse(w, http.StatusBadRequest, fmt.Errorf("token is invalid: %v", err))
 		return
 	}
 
-	valid, apiSession, err := al.validateBearerToken(req.Context(), token, req.URL.Path, req.Method)
+	if al.bannedUsers.IsBanned(tokenCtx.AppToken.Username) {
```
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api_handler_logout.go]: i think @thorstenkramm has a TODO somewhere suggesting that the banning could be moved to middleware? agreed that the checks here could be elsewhere.

File: server/api/session/session.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api/session/session.go]: fyi, i've done a PoC with the `sessionID` added to the app claims in the JWT and tokens removed from the db and the session query json responses. seems to be working fine. all tests are passing and a basic manual test seems to ok. i'd suggest we go ahead with it, unless @thorstenkramm can think of edge cases where it won't or doesn't work?

File: server/api/session/cache.go (hunk 4 above)
Comment (morestatic) [reply server/api/session/cache.go:132]: see comment about PoC for `sessionID` only `api_sessions.db`. for deleting all by `username` we'd still need to loop but for a single session we could just delete and return.

File: server/api_handler_users.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (thorstenkramm) [reply server/api_handler_users.go:189]: @domenipavec @morestatic I would prefer the current implementation. UserID is the ID of the user whose session is destroyed. Storing the session ID is not very helpful. AFAIK, the session ID is not stored elsewhere, so there is no link to an external data point that would add more context to the log entry.

File: server/api_handler_users.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (thorstenkramm) [reply server/api_handler_users.go:212]: @morestatic @domenipavec Why not introduce `ApplicationAuthAPISessions` (Plural)
I think it is precise enough to log if a single or all sessions have been deleted. Logging each session deletion individually creates too much noise on the audit log.  We can add such verbose logging on user request when we know a use case.

**Review by domenipavec — APPROVED** [review]
Please use `ApplicationAuthAPISessions` (Plural) for bulk deletion as suggested by Thorsten. Otherwise looks good.
If you decide to put the code for not storing tokens in this PR, please re-request my review.

File: server/api_handler_users.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api_handler_users.go:212]: added `ApplicationAuthAPISessions`. done.

File: server/api/session/session.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (morestatic) [reply server/api/session/session.go]: ok. i've now included the refactoring to session ids vs tokens as the primary key for the `api_sessions.db`, including removal of the tokens from the db and responses. commit pushed.

File: server/api/session/cache.go (hunk 4 above)
Comment (morestatic) [reply server/api/session/cache.go:138]: now using `sessionID`. see comment above.
